src/common/com/intellij/plugins/haxe/model/type/HaxeTypeCompatible.java

```diff
@@ -144,6 +154,39 @@ static private boolean canAssignToFromFunction(
     return to.returnValue != null && (to.returnValue.isVoid() || to.returnValue.canAssign(from.returnValue));
   }
 
+  private static boolean hasAbstractFunctionTypeCast(SpecificTypeReference reference, boolean isFrom) {
+    if (reference instanceof SpecificHaxeClassReference) {
+      HaxeClass haxeClass = ((SpecificHaxeClassReference)reference).getHaxeClass();
+      if (haxeClass instanceof HaxeAbstractClassDeclaration) {
+        HaxeAbstractClassDeclaration abstractClass = (HaxeAbstractClassDeclaration)haxeClass;
+        if (!abstractClass.getFunctionTypeList().isEmpty()) {
+          for (int i = 0; i < abstractClass.getFunctionTypeList().size(); i++) {
+            if (abstractClass.getIdentifierList().get(i).getText().equals(isFrom ? "to" : "from")) {
+              return true;
+            }
+          }
+        }
+      }
+    }
+    return false;
+  }
+
+  private static SpecificFunctionReference asAbstractFunctionType(SpecificTypeReference reference, boolean isFrom) {
+    SpecificHaxeClassReference classReference = (SpecificHaxeClassReference)reference;
+    HaxeAbstractClassDeclaration abstractClass = (HaxeAbstractClassDeclaration)classReference.getHaxeClass();
+    if (!abstractClass.getFunctionTypeList().isEmpty()) {
+      for (int i = 0; i < abstractClass.getFunctionTypeList().size(); i++) {
+        if (abstractClass.getIdentifierList().get(i).getText().equals(isFrom ? "to" : "from")) {
```

Comment:
NPE on getText() here, too.

Reply:
getText() says @NonNls and the item is from a collection (abstractClass.getIdentifierList().get(i)) så i am not sure how it can be null. but the more i look at it the more i want to change the BNF so that i dont have to do a string compare here

Reply:
`@NonNls` means "Doesn't get translated into other languages".  It is not the same as `@NotNull`. 

I'm with you on letting the parser do the "to" and "from" parsing.  Make a new rule something like:
```bnf
fromKeyword ::= "from"  (can't just make it a token, unfortunately because that's allowed in other places.)
toKeyword ::= "to"
private convertableKeyword ::=  toKeyword | fromKeyword
abstractConvertable ::= convertableKeyword type
```
and use abstractConvertable in abstractClassDeclaration:
```bnf
abstractClassDeclaration ::= privateKeyWord? abstractClassType componentName genericParam? underlyingType? abstractConvertable* abstractBody
```

A year or three ago, I had a rule in there to deal with the to/from parsing, but I never finished that task, so I had left it commented out.  Eventually, somebody else removed it.

Reply:
i have pushed some changes that should handle this as psiElement types now instead of  checking text, let me know what you think